waspc/examples/todoApp/ext/pages/Task.js

```diff
@@ -2,19 +2,44 @@ import React from 'react'
 import { Link } from 'react-router-dom'
 
 import { useQuery } from '@wasp/queries'
+import { useAction } from '@wasp/actions'
 import updateTaskIsDone from '@wasp/actions/updateTaskIsDone'
 import getTask from '@wasp/queries/getTask.js'
+import getTasks from '@wasp/queries/getTasks.js'
 
 const Todo = (props) => {
   const taskId = parseInt(props.match.params.id)
   const { data: task, isFetching, error } = useQuery(getTask, { id: taskId })
 
+  const updateTaskIsDoneAction = useAction(updateTaskIsDone, {
+    optimisticUpdates: [
+      {
+        getQuery: () => [getTask, { id: taskId }],
+        // This query's cache should should never be emtpy
+        updateQuery: ({ isDone }, oldTask) => ({ ...oldTask, isDone }),
+      },
+      {
+        getQuery: () => [getTasks],
+        updateQuery: (updatedTask, oldTasks) => {
+          if (oldTasks === undefined) {
+            // cache is empty
+            return [updatedTask]
+          } else {
+            return oldTasks.map(task => 
+              task.id === updatedTask.id ? { ...task, ...updatedTask } : task
+            )
```

Comment:
Regardless of the choices we make on how to implement optimistic updates (hook vs no-hook, counter vs applyting updates, etc), we'll still have a flickering update problem when it comes to queries that aren't mounted at the moment the optimistic update is performed.

Here's a demo of how it looks. In this example, I'm optimistically updating both `get-tasks` and `get-task` (as is the case in the code). After I leave the Task page and get to the dashboard, React Query automatically invalidates `get-tasks`, causing a refetch of the not-yet-updated state. Finally, once the final response arrives 3 seconds later, we see an extra update:

https://user-images.githubusercontent.com/22752307/177990196-e2ef7533-1357-4620-b026-a035a2f1358e.mp4



By default, React Query refreshes every query whenever a user mounts a component that uses it. This always happens<sup>1</sup> and is not controlled by Wasp. We can stop our own (explicit) invalidations through various mechanisms (i.e., the counter), but we can't stop their (implicit) invalidations without affecting global behavior unrelated to optimistic updates.

The same problem happens when the window is refocused<sup>2</sup> (although this rarely happens outside of development). Here's the demo:

https://user-images.githubusercontent.com/22752307/177991728-6dd6c124-1419-49e3-8444-156377e68997.mp4



[1] Unless the developer configures the query client with  `refetchOnMount: false`, but this does more damage than it helps
[2] Unless the developer configures the query client with `refetchOnWindowFocus: false`, but they probably don't want this behavior either

Reply:
Man, you are uncovering all these problems early (which is great), but I never really imagined how complex all of this was lol. I think Meteor got around most of this by being able to ship Mongo updates but they could rely on the local state and only had to unwind on error, not after competing completions. But overall, good find!

Would it make sense for us to expose some config option where, if enabled, we can overlay some loading component instead of the real one anytime we have optimistic updates pending for these two scenarios? Prob a bad idea/not possible, but just brainstorming. :D